crates/adapters/src/controller/mod.rs

```diff
@@ -1051,56 +1070,63 @@ struct StepTrigger {
 
     /// Time between clock ticks.
     clock_resolution: Option<Duration>,
+
+    /// Time between automatic checkpoints.
+    checkpoint_interval: Option<Duration>,
 }
 
-/// When the next step will definitely trigger (if not before).
-enum Wait {
-    /// At time `.0`.
-    Time(Instant),
+/// Action for the controller to take.
+enum Action {
+    /// Park until time `.0`, or forever if `None`.
+    Park(Option<Instant>),
 
-    /// No step is scheduled.
-    Forever,
-}
+    /// Write a checkpoint.
+    Checkpoint,
 
-impl Wait {
-    // Park until the next step will trigger (or until we're unparked).
-    fn park(self, parker: &Parker) {
-        match self {
-            Wait::Time(t) => parker.park_deadline(t),
-            Wait::Forever => parker.park(),
-        }
-    }
+    /// Step the circuit.
+    Step,
 }
 
 impl StepTrigger {
     /// Returns a new [StepTrigger].
     fn new(controller: Arc<ControllerInner>) -> Self {
-        let config = &controller.status.pipeline_config;
-        let max_buffering_delay = Duration::from_micros(config.global.max_buffering_delay_usecs);
-        let min_batch_size_records = config.global.min_batch_size_records;
-        let clock_resolution = config
-            .global
-            .clock_resolution_usecs
-            .map(Duration::from_micros);
+        let config = &controller.status.pipeline_config.global;
+        let max_buffering_delay = Duration::from_micros(config.max_buffering_delay_usecs);
+        let min_batch_size_records = config.min_batch_size_records;
+        let clock_resolution = config.clock_resolution_usecs.map(Duration::from_micros);
+        let checkpoint_interval =
+            config
+                .fault_tolerance
+                .and_then(|ft| match ft.checkpoint_interval_secs {
+                    0 => None,
+                    secs => Some(Duration::from_secs(secs)),
+                });
         Self {
             controller,
             tick: clock_resolution.map(|delay| Instant::now() + delay),
             buffer_timeout: None,
             max_buffering_delay,
             min_batch_size_records,
             clock_resolution,
+            checkpoint_interval,
         }
     }
 
     /// Determines when to trigger the next step, given whether we're currently
     /// `replaying` and whether the pipeline is currently `running`.  Returns
     /// `None` to trigger a step right away, and otherwise how long to wait.
-    fn trigger(&mut self, replaying: bool, running: bool) -> Option<Wait> {
+    fn trigger(&mut self, last_checkpoint: Instant, replaying: bool, running: bool) -> Action {
         let buffered_records = self.controller.status.num_buffered_input_records();
 
         // `self.tick` but `None` if we're not running.
         let tick = running.then_some(self.tick).flatten();
 
+        if let Some(checkpoint_interval) = self.checkpoint_interval {
```

Comment:
We could do a lot of checkpoints, yes, but I don't think we'll end up doing only checkpoints because the interval is denominated in seconds and we set `last_checkpoint` after we take the checkpoint, see this:

```rust
        // Update the last checkpoint time *after* executing the checkpoint, so
        // that if the checkpoint takes a long time and we have a short
        // checkpoint interval, we get to do other work too.
        self.last_checkpoint = Instant::now();
```

If this turns out to be a problem, then we can take countermeasures, such as increasing the minimum interval.